### DISCUSSION TITLE
HttpClient + Aff<T> instead of Task<T>

### DISCUSSION THREAD

**Discussion by rmbrunet** [louthy/language-ext#1088] [Q&A] [answered]
Hello, 

Trying to understand (better or at all) the usage of **Aff**.

We usually model the interaction with the outside world with **Task<T>**. **Aff<T>** is a better description. Trying with **HttpClient** with the  purpose of: a) wrap the possible **!HttpResponseMessage.IsSuccessStatusCode** in an **Error**, and b) enforce the non-nullability of the return value. 

It seems with Aff<T>() the options (probably by design) are: return T or throw an exception. No possibility of returning Error, Fail*, etc..  This force composability to achieve the goals as stated before. Can someone please check the code below and comment how far from the right/better way it is?

Greatly appreciate it!

Note: avoiding the Runtime for now.

``` C#
public class UserService : IUserService
{
    private readonly HttpClient _client;

    public UserService(HttpClient client)
    {
        _client = client;
    }

    public Aff<IEnumerable<User>> GetUsers()
        => Get<IEnumerable<User>>(_client, "users");

    public Aff<User> GetUser(int id)
        => Get<User>(_client, $"users/{id}");

    private static Aff<T> Get<T>(HttpClient client, string url)
    => from a in GetHttpResponse(client, url)
       from b in CheckHttpResponseError(a)
       from c in Get<T>(b)
       from d in Optional(c).ToEff()
       select d;

    private static Aff<T?> Get<T>(HttpResponseMessage response)
             => Aff(async () => await response.Content.ReadFromJsonAsync<T>());

    private static Aff<HttpResponseMessage> GetHttpResponse(HttpClient client, string url)
        => Aff<HttpResponseMessage>(async () => await client.GetAsync(url));

    private static Eff<HttpResponseMessage> CheckHttpResponseError(HttpResponseMessage t)
        => t switch
        {
            { IsSuccessStatusCode: true } => SuccessEff(t),
            _ => FailEff<HttpResponseMessage>(Error.New((int)t.StatusCode, t.ReasonPhrase, thisException: new HttpRequestException(t.ReasonPhrase, null, t.StatusCode)))
        };
}
```

**Comment by louthy** [comment, marked as answer]
> return T or throw an exception

No, it returns `T` or `Error` (which allows for exceptional and non-exceptional errors).  See the latest beta release for details on [using and extending the `Error` type](https://github.com/louthy/language-ext/releases/tag/v4.3.0-beta).  You can also use `cancel<RT>()` to exit an `Aff` expression.

Your code above is fine.  I would personally prefer not to use a stateful class and instead use the _runtime_ capability of the effect system, but that's really your design choice.

If you are interested in the runtime approach, then try this:

Define an interface that represents HTTP client IO:

```c#
/// <summary>
/// Interface to the HttpClient 
/// </summary>
public interface HttpClientIO
{
    Task<HttpResponseMessage> GetAsync(string? requestUri);
}
```
This abstracts awat from `HttpClient` and allows for injectable behaviours (for unit testing and the like).

Then define a `HasHttpClient` trait interface:
```c#
/// <summary>
/// Trait that we use when building a runtime to state 'we have support for HttpClient'
/// </summary>
/// <typeparam name="RT">The runtime</typeparam>
public interface HasHttpClient<RT> where RT : struct, HasCancel<RT>, HasHttpClient<RT>
{
    /// <summary>
    /// Access to the HttpClientIO interface
    /// </summary>
    Eff<RT, HttpClientIO> HttpClientEff { get; }
}
```
This allows the runtime to easily hook into pre-built `HttpClientIO` implementations.

Now create a 'live' implementation:
```c#
/// <summary>
/// The real implementation of HttpClientIO.  We could have other implementations for unit-testing
/// </summary>
/// <param name="Client">Shared client to use</param>
public record HttpClientLive(HttpClient Client) : HttpClientIO
{
    public Task<HttpResponseMessage> GetAsync(string? requestUri) =>
        Client.GetAsync(requestUri);
}
```
Obviously building a unit-test version of this shouldn't be too hard.

Now create an environment for the runtime:

```c#
/// <summary>
/// Environment variables used in the runtime.  This can be seen as configuration,
/// but also contextual state.
/// </summary>
/// <remarks>We keep this in a separate type because the runtime is a struct.  This
/// stops the struct copying getting to cumbersome.</remarks>
/// <param name="HttpClient">Shared HTTP client </param>
/// <param name="CancellationToken">Concellation token</param>
/// <param name="CancellationTokenSource">Concellation token source</param>
public record RuntimeEnv(
    HttpClient HttpClient,
    CancellationToken CancellationToken, 
    CancellationTokenSource CancellationTokenSource) : IDisposable
{
    public static RuntimeEnv New(HttpClient httpClient)
    {
        var src = new CancellationTokenSource();
        return new RuntimeEnv(httpClient, src.Token, src);
    }

    /// <summary>
    /// Create a local cancellation context
    /// </summary>
    public RuntimeEnv LocalCancel
    {
        get
        {
            var src = new CancellationTokenSource();
            return this with {CancellationToken = src.Token, CancellationTokenSource = src};
        }
    }

    public void Dispose()
    {
        HttpClient?.Dispose();
        CancellationTokenSource.Dispose();
    }
}
```
This is where configuration, environment, and contextual state lives.  It will hold the one and only `HttpClient` for the application.

Now define the runtime itself.
```c#
/// <summary>
/// The runtime itself, we only implement two traits: HasCancel and HasHttpClient
/// Other traits are available!
/// </summary>
public readonly struct Runtime : HasCancel<Runtime>, HasHttpClient<Runtime>, IDisposable
{
    readonly RuntimeEnv env;

    /// <summary>
    /// Ctor
    /// </summary>
    public Runtime(RuntimeEnv env) =>
        this.env = env;        
    
    /// <summary>
    /// Access to the environment
    /// </summary>
    /// <exception cref="Exception"></exception>
    public RuntimeEnv Env => 
        env ?? throw new Exception("Runtime not initialised correctly");

    /// <summary>
    /// Access to the real HttpClientIO implementation
    /// </summary>
    public Eff<Runtime, HttpClientIO> HttpClientEff =>
        Eff<Runtime, HttpClientIO>(rt => new HttpClientLive(rt.Env.HttpClient));

    /// <summary>
    /// Create a local cancellation context
    /// </summary>
    public Runtime LocalCancel =>
        new(Env.LocalCancel);

    /// <summary>
    /// Cancellation token
    /// </summary>
    public CancellationToken CancellationToken =>
        env.CancellationToken;
    
    /// <summary>
    /// Cancellation token source
    /// </summary>
    public CancellationTokenSource CancellationTokenSource =>
        env.CancellationTokenSource;

    public void Dispose() => 
        env?.Dispose();
}
```
A runtime implements a number of `Has*` traits: by extending the set of traits implemented, you extend the set of sanctioned IO and global environment state values available to your application.

Now let's create a specific error type for HTTP errors:

```c#
public record HttpError(string Message, int Code) : Expected(Message, Code)
{
    public HttpError(HttpResponseMessage msg) : this(msg.ReasonPhrase ?? "HTTP error", (int)msg.StatusCode) { }
}
```
We derive it from `Expected` because these aren't exceptional errors.

Now we can create static functions that work with the runtime:

```c#
/// <summary>
/// Static access to the HTTP Client IO
/// </summary>
public static class Http<RT> 
    where RT : struct, HasHttpClient<RT>, HasCancel<RT>
{
    public static Aff<RT, A> get<A>(string? requestUri) =>
        from io in default(RT).HttpClientEff
        from rm in AffMaybe(async () =>
                        await io.GetAsync(requestUri) switch
                        {
                            {IsSuccessStatusCode: true} m => FinSucc(m),
                            var m                         => FinFail<HttpResponseMessage>(new HttpError(m))
                        })
        from js in Aff(async () => await rm.Content.ReadFromJsonAsync<A>())
        select js;
}
```

Notice how it wraps up error handling and deserialisation.  

All of what I've written above, you do once (for any kind of IO your application performs).  So, even though it seems a bit heavyweight, it's a onetime thing - and then you get really slick results for your application code:

```c#
public static class UserService<RT>
    where RT : struct, HasHttpClient<RT>, HasCancel<RT>
{
    public static Aff<IEnumerable<User>> Users = 
        Http<RT>.get<IEnumerable<User>>("users");

    public static Aff<User> GetUser(int id) =>
        Http<RT>.get<User>($"users/{id}");
}
```

`UserService<RT>` is now supports injectable IO for all of its behaviours.  It also declares, completely, all of the IO operations it performs (great for audit and communicating to other devs what the code does).

As I said earlier, your code is totally fine (although you might want to extend the `Error` type you use). This runtime approach is the ultimate though in terms of power.  [Full discussion here](https://github.com/louthy/language-ext/wiki/How-to-deal-with-side-effects).

**Reply by rmbrunet** [reply]
Thanks Paul! I lot of reading and thinking to do. More so, considering that the correlation between the amount of thinking and the resulting understanding is somewhat low.

I was trying to do (to map to an Expected Error) something like:

``` C#
private Aff<HttpResponseMessage> GetAff(HttpClient client, string url)
        => Aff(async () => 
        {
            var response = await client.GetAsync(url);
            return response.IsSuccessStatusCode
                ? response
                : Error.New((int)response.StatusCode, response.ReasonPhrase);
        });
```
that obviously does not work considering that the delegate expects a ValueTask<T> as the return value. 

Definitely interested in the Runtime path but, at the same time, trying to determine a Minimum Viable Set of Skills: the minimum necessary knowledge for a developer without previous experience (in FP or LanguageExt) to be able to be productive under supervision. Then, handling Exceptions and nulls with Eff/Aff seems like a good starting point. 

Thanks again. Regards.

**Reply by louthy** [reply]
Ah, you can call AffMaybe(…) instead of Aff(…).  The result type there is 

    ValueTask<Fin<A>>

**Reply by rmbrunet** [reply]
Oh... this is good! 

``` C#
public class UserService : IUserService
{
    private readonly HttpClient _client;

    public UserService(HttpClient client)
    {
        _client = client;
    }

    public Aff<IEnumerable<User>> GetUsers()
        => Get<IEnumerable<User>>(_client, "users");

    public Aff<User> GetUser(int id)
        => Get<User>(_client, $"users/{id}");

    private static Aff<T> Get<T>(HttpClient client, string url)
        => from a in GetHttpResponse(client, url)
           from b in Get<T>(a)
           select b;

    private static Aff<T> Get<T>(HttpResponseMessage response)
        => AffMaybe<T>(async () =>
        {
            T? t = await response.Content.ReadFromJsonAsync<T>();
            return Optional(t).ToFin();
        });

    private static Aff<HttpResponseMessage> GetHttpResponse(HttpClient client, string url)
        => AffMaybe<HttpResponseMessage>(async () =>
        {
            var response = await client.GetAsync(url);
            return response.IsSuccessStatusCode
                ? response
                : Error.New((int)response.StatusCode, response.ReasonPhrase);
        });
}
```
Going to put this in a repo (similar to the Contoso sample) and think about the runtime. 

Thanks a lot!

**Reply by rmbrunet** [reply]
Created a simple POC to encapsulate what may be the simplest, limited use of LanguageExt in an ASPNET.CORE Web api application. It may be useful to someone.

https://github.com/rmbrunet/Functoso

(Thought for a second to name it Afftoso)

It resulted in a bunch of side effects injected into the chain. Not only HttpClient, but also IValidator from FluentValidation and IMapper from Automapper. And some are still missing (ILogger). Some of these dependency could be removed and implemented inside de chain with Pure functions but then, should we renounce to use these packages (impractical imo), or should be included in the Runtime (seems like a lot of work - But I may be wrong).  Something is clearer now but more questions remain - as always.
